File: Medium/1584. Min Cost to Connect All Points/1584. Min Cost to Connect All Points.py
import heapq


class Solution:
    def minCostConnectPoints(self, points: List[List[int]]) -> int:
        n = len(points)
        visited = [False] * n
        min_costs = [float("inf")] * n
        min_costs[0] = 0
        min_cost = 0

        def manhattan_distance(point1, point2):
            return abs(point1[0] - point2[0]) + abs(point1[1] - point2[1])

        min_heap = [(0, 0)]  # (cost, point_index)

        while min_heap:
            cost, u = heapq.heappop(min_heap)
            if visited[u]:
                continue
            visited[u] = True
            min_cost += cost

            for v in range(n):
                if not visited[v]:
                    edge_cost = manhattan_distance(points[u], points[v])
                    if edge_cost < min_costs[v]:
                        min_costs[v] = edge_cost
                        heapq.heappush(min_heap, (edge_cost, v))

        return min_cost

File: Medium/1584. Min Cost to Connect All Points/1584. Min Cost to Connect All Points.rs
use std::cmp::Reverse;
use std::collections::BinaryHeap;

impl Solution {
    pub fn min_cost_connect_points(points: Vec<Vec<i32>>) -> i32 {
        let n = points.len();
        let mut visited = vec![false; n];
        let mut min_costs = vec![i32::MAX; n];
        min_costs[0] = 0;
        let mut min_cost = 0;
        let mut min_heap = BinaryHeap::new();

        // Start from the first point
        min_heap.push(Reverse((0, 0))); // (cost, point_index)

        while let Some(Reverse((cost, u))) = min_heap.pop() {
            if visited[u] {
                continue;
            }
            visited[u] = true;
            min_cost += cost;

            for v in 0..n {
                if !visited[v] {
                    let edge_cost =
                        (points[u][0] - points[v][0]).abs() + (points[u][1] - points[v][1]).abs();
                    if edge_cost < min_costs[v] {
                        min_costs[v] = edge_cost;
                        min_heap.push(Reverse((edge_cost, v)));
                    }
                }
            }
        }

        min_cost
    }
}
